File: Agora-Signaling-Tutorial-iOS-Objective-C/Agora-Signaling-Tutorial/Controllers/ChatViewController.m
//
//  ChatViewController.m
//  Agora-Signaling-Tutorial
//
//  Created by Zhangji on 2018/11/16.
//  Copyright © 2018 Agora. All rights reserved.
//

#import "ChatViewController.h"
#import "AgoraSignal.h"
#import "Message.h"
#import "ChatRoomViewController.h"

@interface ChatViewController ()
@property (weak, nonatomic) IBOutlet UITextField *chatAccountTextField;
@end

@implementation ChatViewController
- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.navigationItem.backBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"" style:UIBarButtonItemStylePlain target:nil action:nil];
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"logout"] style:UIBarButtonItemStylePlain target:self action:@selector(logout)];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self addAgoraSignalBlock];
}

- (void)logout {
    [[AgoraSignal sharedKit] logout];
}

- (IBAction)chatButtonClicked:(UIButton *)sender {
    NSString *chatAccount = self.chatAccountTextField.text;
    if (!chatAccount.length) {
        return;
    }
    if (![self checkString:chatAccount]) {
        return;
    }
    [self performSegueWithIdentifier:@"ShowChatRoom" sender:chatAccount];
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [self.view endEditing:YES];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    ChatRoomViewController *chatRoomVC = (ChatRoomViewController *)(segue.destinationViewController);
    NSString *account = sender;
    chatRoomVC.account = account;
    MessageList *messageList = [MessageCenter sharedCenter].chatMessages[account];
    if (messageList) {
        chatRoomVC.messageList = messageList;
    } else {
        chatRoomVC.messageList.identifier = account;
    }
}

- (void)addAgoraSignalBlock {
    __weak typeof(self) weakSelf = self;
    [AgoraSignal sharedKit].onLogout = ^(AgoraEcode ecode) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf dismissViewControllerAnimated:YES completion:nil];
        });
    };
    
    [AgoraSignal sharedKit].onMessageInstantReceive = ^(NSString *account, uint32_t uid, NSString *msg) {
        NSMutableDictionary *chatMessages = [MessageCenter sharedCenter].chatMessages;
        MessageList *messageList = chatMessages[account];
        
        if (!messageList) {
            chatMessages[account] = [[MessageList alloc] initWithIdentifier:account];
        }
        Message *message = [[Message alloc] initWithAccount:account message:msg];
        
        [messageList.list addObject:message];
    };
}

- (BOOL)checkString:(NSString *)string {
    if (!string.length || string.length > 128) {
        return NO;
    }
    
    if ([string containsString:@" "]) {
        return NO;
    }
    
    return YES;
}

- (void)alertString:(NSString *)string {
    if (!string.length) {
        return;
    }
    
    __weak typeof(self) weakSelf = self;
    dispatch_async(dispatch_get_main_queue(), ^{
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:nil message:string preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction *action = [UIAlertAction actionWithTitle:@"Ok" style:UIAlertActionStyleCancel handler:nil];
        [alert addAction:action];
        [weakSelf presentViewController:alert animated:YES completion:nil];
    });
}
@end

File: Agora-Signaling-Tutorial-iOS-Swift/Agora-Signaling-Tutorial/Controllers/ChatViewController.swift
//
//  ChatViewController.swift
//  Agora-Signaling-Tutorial
//
//  Created by ZhangJi on 04/12/2017.
//  Copyright © 2017 ZhangJi. All rights reserved.
//

import UIKit

class ChatViewController: UIViewController {
    @IBOutlet weak var chatAccountTextField: UITextField!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.navigationItem.backBarButtonItem = UIBarButtonItem(title: "", style: .plain, target: nil, action: nil)
        self.navigationItem.rightBarButtonItem = UIBarButtonItem(image: #imageLiteral(resourceName: "logout"), style: .plain, target: self, action: #selector(logout))
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        addAgoraSignalBlock()
    }
    
    @objc func logout() {
        AgoraSignal.Kit.logout()
    }
    
    @IBAction func chatButtonClicked(_ sender: UIButton) {
        guard let chatAccount = chatAccountTextField.text else {
            return
        }
        if !check(String: chatAccount) {
            return
        }
        self.performSegue(withIdentifier: "ShowChatRoom", sender: chatAccount)
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        view.endEditing(true)
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        let chatRoomVC = segue.destination as! ChatRoomViewController
        chatRoomVC.account = sender as! String
        if chatMessages[chatRoomVC.account] != nil {
            chatRoomVC.messageList = chatMessages[chatRoomVC.account]!
        } else {
            chatRoomVC.messageList.identifier = chatRoomVC.account
        }
    }
    
    func addAgoraSignalBlock() {
        AgoraSignal.Kit.onLogout = { [weak self] (ecode) -> () in
            DispatchQueue.main.async(execute: {
                self?.dismiss(animated: true, completion: nil)
            })
        }
        
        AgoraSignal.Kit.onMessageInstantReceive = { (account, uid, msg) -> () in
            if chatMessages[account!] == nil {
                let message = Message(account: account, message: msg)
                var messagelist = [Message]()
                messagelist.append(message)
                let chatMsg = MessageList(identifier: account, list: messagelist)
                chatMessages[account!] = chatMsg
                return
            }
            let message = Message(account: account, message: msg)
            chatMessages[account!]?.list.append(message)
        }
    }
}

extension ChatViewController {
    func check(String: String) -> Bool {
        if String.isEmpty {
            alert(string: "The account is empty !")
            return false
        }
        if String.count > 128 {
            alert(string: "The accout is longer than 128 !")
            return false
        }
        if String.contains(" ") {
            alert(string: "The accout contains space !")
            return false
        }
        return true
    }
    
    func alert(string: String) {
        guard !string.isEmpty else {
            return
        }
        
        DispatchQueue.main.async(execute: {
            let alert = UIAlertController(title: nil, message: string, preferredStyle: .alert)
            alert.addAction(UIAlertAction(title: "Ok", style: .cancel, handler: nil))
            self.present(alert, animated: true, completion: nil)
        })
    }
}
